### PR TITLE
Parsed Declarative Web Push JSON should be remembered and round trip throughout various forms of IPC

### PR DESCRIPTION
**Pull request by beidson**
#### 5043150257293a0e89dd2166d45592e9ac1cc379
<pre>
Parsed Declarative Web Push JSON should be remembered and round trip throughout various forms of IPC
<a href="https://bugs.webkit.org/show_bug.cgi?id=261632">https://bugs.webkit.org/show_bug.cgi?id=261632</a>
rdar://115585968

Reviewed by Alex Christensen.

webpushd knows how to validate Declarative Notification JSON into a NotificationPayload

But now:
- Once validated, it saves the NotificationPayload into the push message that will be fetched by the client later
- NotificationPayload and related all know how to serialize, both in the CoreIPC sense and Obj-C SPI client sense
- Tests validate that all that serialization happens.

* Source/WebCore/Modules/notifications/NotificationDirection.h:
* Source/WebCore/Modules/notifications/NotificationJSONParser.cpp:
(WebCore::NotificationJSONParser::parseNotificationPayload):
(WebCore::NotificationJSONParser::parseNotificationOptions):
* Source/WebCore/Modules/notifications/NotificationJSONParser.h:
* Source/WebCore/Modules/notifications/NotificationOptionsPayload.h: Copied from Source/WebCore/Modules/notifications/NotificationPayload.h.
* Source/WebCore/Modules/notifications/NotificationOptionsPayloadCocoa.mm: Added.
(WebCore::NotificationOptionsPayload::fromDictionary):
(WebCore::NotificationOptionsPayload::dictionaryRepresentation const):
* Source/WebCore/Modules/notifications/NotificationPayload.h:
* Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm: Added.
(WebCore::NotificationPayload::fromDictionary):
(WebCore::NotificationPayload::dictionaryRepresentation const):
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/IPC/DaemonCoders.cpp:
* Source/WebKit/Platform/IPC/DaemonCoders.h:
(WebKit::Daemon::Coder&lt;WebPushD::PushMessageForTesting&gt;::encode): Deleted.
(WebKit::Daemon::Coder&lt;WebPushD::PushMessageForTesting&gt;::decode): Deleted.
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedEnum.is_webkit_platform):
(generate_impl):
(parse_serialized_types):
* Source/WebKit/Shared/Cocoa/WebPushMessageCocoa.mm:
(WebKit::WebPushMessage::fromDictionary):
(WebKit::WebPushMessage::toDictionary const):
* Source/WebKit/Shared/PushMessageForTesting.h:
(WebKit::WebPushD::PushMessageForTesting::encode const): Deleted.
(WebKit::WebPushD::PushMessageForTesting::decode): Deleted.
* Source/WebKit/Shared/PushMessageForTesting.serialization.in: Copied from Source/WebKit/Shared/WebPushMessage.serialization.in.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebPushMessage.h:
* Source/WebKit/Shared/WebPushMessage.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/webpushd/PushService.mm:
(WebPushD::PushService::didReceivePushMessage):
* Source/WebKit/webpushd/WebPushDaemon.h:
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::injectPushMessageForTesting):
(WebPushD::WebPushDaemon::getPendingPushMessages):
* Source/WebKit/webpushd/webpushtool/WebPushToolMain.mm:
(pushMessageFromArguments):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:

Canonical link: <a href="https://commits.webkit.org/268052@main">https://commits.webkit.org/268052@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7453a3eb048bdf6baa0e3a4172eab485c2136b50

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18526 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19465 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17328 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18721 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22173 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19005 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19204 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18750 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/18925 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16136 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21258 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/18691 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16147 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16898 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23351 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/17166 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17068 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21246 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17645 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14954 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16705 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4403 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21069 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17487 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        rawAppBadge = [appBadge unsignedLongLongValue];
+    }
+
+    NSDictionary *options = dictionary[WebNotificationOptionsKey];
+    if (!options)
+        return std::nullopt;
+
+    std::optional<NotificationOptionsPayload> rawOptions;
+    if (![options isKindOfClass:[NSNull class]]) {
+        if (![options isKindOfClass:[NSDictionary class]])
```
Comment (achristensen07) [inline Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm]: Move to inside NotificationOptionsPayload::fromDictionary?  That would make it more robust.

File: Source/WebKit/Scripts/generate-serializers.py [hunk 2]
```diff
@@ -167,6 +167,8 @@ def is_option_set(self):
     def is_nested(self):
         return 'Nested' in self.attributes
 
+    def is_webkit_platform(self):
```
Comment (achristensen07) [inline Source/WebKit/Scripts/generate-serializers.py:170]: Having EnumTraits and serialization.in will cause linker issues for MSVC.  For now let's just remove all this file's changes, and switch NotificationPayload to only having EnumTraits until we can figure out what to do in a separate PR.

**Review by achristensen07 — APPROVED** [review]

File: Source/WebKit/Scripts/generate-serializers.py (hunk 2 above)
Comment (beidson) [reply Source/WebKit/Scripts/generate-serializers.py:170]: Would I expect to see that breakage on an EWS? Currently looks like it's fine?

Comment (beidson) [reply Source/WebKit/Scripts/generate-serializers.py:170]: EWS passed, but I'll make that change anyways

File: Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+static NSString * const WebNotificationDefaultActionKey = @"WebNotificationDefaultActionKey";
+static NSString * const WebNotificationTitleKey = @"WebNotificationTitleKey";
+static NSString * const WebNotificationAppBadgeKey = @"WebNotificationAppBadgeKey";
+static NSString * const WebNotificationOptionsKey = @"WebNotificationOptionsKey";
+
+namespace WebCore {
+
+std::optional<NotificationPayload> NotificationPayload::fromDictionary(NSDictionary *dictionary)
+{
+    NSURL *defaultAction = dictionary[WebNotificationDefaultActionKey];
```
Comment (achristensen07) [inline Source/WebCore/Modules/notifications/NotificationPayloadCocoa.mm:40]: This could also check if its a dictionary.

File: Source/WebKit/Scripts/generate-serializers.py (hunk 2 above)
Comment (achristensen07) [reply Source/WebKit/Scripts/generate-serializers.py:170]: These changes to this file are no longer used or tested.
